File: packages/ai/agents/math/format.ts
import type { MathData } from "@repo/math/schema/data";
import type { MathResult } from "@repo/math/schema/result";
import type {
  MathExpression,
  MathItem,
  MathStep,
} from "@repo/math/schema/shared";
import dedent from "dedent";

/** Formats deterministic checked work as model-readable markdown. */
export function formatMathData(data: MathData, diagnostic?: string) {
  if (data.status === "loading") {
    return "Checked math work is loading.";
  }

  if (data.status === "error") {
    return dedent(`
      # Checked Math Work

      - Status: error
      - Operation: ${data.kind}
      - Evidence scope: unavailable deterministic derivation; do not describe the result as checked or proven
      - Error code: ${data.error}
      ${formatDiagnostic(diagnostic)}
    `);
  }

  return dedent(`
    # Checked Math Work

    - Status: ${data.status}
      - Operation: ${data.result.operation}
      - Step status: ${data.result.stepStatus}
      - Evidence scope: ${formatEvidenceScope(data.result)}
      - Reason: ${data.result.reason}
      - Primary: ${data.result.primary.expression}
    ${formatSecondary(data.result)}
    ${formatSteps(data.result.steps)}
    ${formatItems(data.result.items)}
    ${formatConditions(data.result.conditions)}
  `);
}

/** Formats optional model-only recovery guidance for failed math checks. */
function formatDiagnostic(diagnostic?: string) {
  if (!diagnostic) {
    return "";
  }

  return `- Recovery: ${diagnostic}`;
}

/** Formats the exact verification scope for the assistant-facing evidence. */
function formatEvidenceScope(result: MathResult) {
  if (result.stepStatus === "complete") {
    return "complete deterministic derivation";
  }

  if (result.stepStatus === "partial") {
    if (result.status === "verified") {
      return "verified deterministic result with partial derivation steps";
    }

    return "partial deterministic evidence; do not call the final result fully verified without another complete check";
  }

  if (result.status === "verified") {
    return "verified deterministic result; derivation steps are not included";
  }

  return "unavailable deterministic derivation; say the available evidence is not enough to prove the result";
}

/** Formats the optional second expression in the math result. */
function formatSecondary(result: MathResult) {
  if (result.steps.length > 0) {
    return "";
  }

  if (!result.secondary) {
    return "";
  }

  return `- Secondary: ${result.secondary.expression}`;
}

/** Formats deterministic math derivation steps as model-readable bullets. */
function formatSteps(steps: readonly MathStep[]) {
  if (steps.length === 0) {
    return "";
  }

  return steps
    .map((step) => {
      const relation = step.relation ? ` ${step.relation.expression}` : "";
      const secondary = step.secondary ? ` ${step.secondary.expression}` : "";
      return `- Step (${step.action}): ${step.primary.expression}${relation}${secondary}`;
    })
    .join("\n");
}

/** Formats supporting math rows as model-readable markdown bullets. */
function formatItems(items: readonly MathItem[]) {
  if (items.length === 0) {
    return "";
  }

  return items.map((item) => `- ${item.label}: ${item.value}`).join("\n");
}

/** Formats math conditions such as domain restrictions. */
function formatConditions(conditions: readonly MathExpression[]) {
  if (conditions.length === 0) {
    return "";
  }

  return conditions
    .map((condition) => `- Condition: ${condition.expression}`)
    .join("\n");
}

File: apps/cas/cas/format.py
"""Shared result formatting for CAS responses."""

from collections.abc import Sequence

import sympy as sp

from cas.schema import (
    MathExpression,
    MathItem,
    MathRequest,
    MathResult,
    MathStatus,
    MathStep,
    MathStepStatus,
)


def expression(value: object) -> MathExpression:
    """Render a CAS value as plain text and LaTeX."""
    return MathExpression(expression=str(value), latex=sp.latex(value))


def expression_text(value: str, latex: str | None = None) -> MathExpression:
    """Render expression text that is already display-safe."""
    return MathExpression(expression=value, latex=latex or value)


def math_expression(value: object) -> MathExpression:
    """Keep display-safe expressions intact and render all other CAS values."""
    if isinstance(value, MathExpression):
        return value

    return expression(value)


def item(label: str, value: object) -> MathItem:
    """Render a labeled supporting CAS value."""
    rendered = math_expression(value)

    return MathItem(label=label, value=rendered.expression, latex=rendered.latex)


def step(
    action: str,
    *,
    primary: object,
    relation: MathExpression | None = None,
    secondary: object | None = None,
    items: Sequence[MathItem] | None = None,
) -> MathStep:
    """Build one deterministic math evidence step."""
    return MathStep(
        action=action,
        primary=math_expression(primary),
        relation=relation,
        secondary=math_expression(secondary) if secondary is not None else None,
        items=list(items or []),
    )


def result(
    request: MathRequest,
    *,
    status: MathStatus,
    primary: object,
    reason: str,
    secondary: object | None = None,
    items: Sequence[MathItem] | None = None,
    conditions: Sequence[object] | None = None,
    steps: Sequence[MathStep] | None = None,
    stepStatus: MathStepStatus = "unavailable",
) -> MathResult:
    """Build the stable result envelope shared with the TypeScript client."""
    return MathResult(
        kind=request.operation,
        operation=request.operation,
        status=status,
        input=request,
        primary=math_expression(primary),
        secondary=math_expression(secondary) if secondary is not None else None,
        items=list(items or []),
        conditions=[
            condition
            if isinstance(condition, MathExpression)
            else expression(condition)
            for condition in conditions or []
        ],
        steps=list(steps or []),
        stepStatus=stepStatus,
        reason=reason,
    )


def inconclusive(request: MathRequest, reason: str) -> MathResult:
    """Build an honest inconclusive result for unsupported operations."""
    return result(
        request,
        status="inconclusive",
        primary=request.expression or request.left or request.operation,
        reason=reason,
    )
